### ISSUE TITLE
0.12-beta fails to play anything on kodi 18

### ISSUE THREAD

**Issue by pitsi** [mrgsi/repository.wnt2#4]
As mentioned in the title, wnt2 fails to play anything on kodi 18 (libreelec 9.2.6 x64 on my end) and brings up this error
```
2023-11-23 08:40:17.371 T:139995291510528   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.TypeError'>
                                            Error Contents: 'encoding' is an invalid keyword argument for this function
                                            Traceback (most recent call last):
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/default.py", line 5, in <module>
                                                main()
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/lib/plugin.py", line 1704, in main
                                                globals()[params.get('action', 'actionMenu')](params)
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/lib/plugin.py", line 1384, in actionResolve
                                                recently_watched_add( parent['name'], parent['url'] )
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/lib/recently_watched.py", line 38, in recently_watched_add
                                                data = recently_watched_load()
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/lib/recently_watched.py", line 22, in recently_watched_load
                                                with open( RW_FILE, 'r', encoding='utf8' ) as rw_str:
                                            TypeError: 'encoding' is an invalid keyword argument for this function
                                            -->End of Python script error report<--
2023-11-23 08:40:17.495 T:139995453204608   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.watchnixtoons2/?action=actionResolve&url=https%3A%2F%2Fwww.wcostream.tv%2Ffoodiverse-2021]
```
By "play" I mean the action when clicking on a movie or a series episode, which would bring up the quality dialog and then actually play the stream. All menus and submenus work, episode listings on series work and search works too.
Moreover, although I think I have all my settings at their defaults values, since ever, I tried changing the domain from there but the same happens
```
2023-11-23 08:45:16.881 T:139995283048192   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.TypeError'>
                                            Error Contents: 'encoding' is an invalid keyword argument for this function
                                            Traceback (most recent call last):
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/default.py", line 5, in <module>
                                                main()
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/lib/plugin.py", line 1704, in main
                                                globals()[params.get('action', 'actionMenu')](params)
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/lib/plugin.py", line 1384, in actionResolve
                                                recently_watched_add( parent['name'], parent['url'] )
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/lib/recently_watched.py", line 38, in recently_watched_add
                                                data = recently_watched_load()
                                              File "/storage/.kodi/addons/plugin.video.watchnixtoons2/lib/recently_watched.py", line 22, in recently_watched_load
                                                with open( RW_FILE, 'r', encoding='utf8' ) as rw_str:
                                            TypeError: 'encoding' is an invalid keyword argument for this function
                                            -->End of Python script error report<--
2023-11-23 08:45:17.013 T:139995453204608   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.watchnixtoons2/?action=actionResolve&url=https%3A%2F%2Fwww.wcofun.tv%2Ffoodiverse-2021]
```
I don't know if the issue is related to python2's shortcomings (compared to python3) and right now I can not test it in a python3 compatible version of kodi, i.e on kodi 19 or newer.

I do know that you are not the developer of the addon, but I do not have a reddit account so as to contact the original dev, so this is the only way for someone to forward the issue to him. I have been using wnt2 for more than a year now, only on that kodi 18 installation, I get the zip from the related thread on reddit and install/update it manually, but this is the first time I run into an issue.

Thank you in advance :)

**Comment by pitsi** [comment]
In order to comply with the troubleshooting guidelines from the addon's thread in r/kodi_helpers, I am adding the following

Both sites work fine in a browser. Their players also work.

The addon is on the latest version
```
# cat .kodi/addons/plugin.video.watchnixtoons2/addon.xml | grep vers
ion
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<addon id="plugin.video.watchnixtoons2" name="WatchNixtoons2" version="0.12.0~beta" provider-name="WNT2">
```

Uninstalling and reinstalling did not help.

Switcing to http only did not help either.

**Comment by mrgsi** [comment]
I haven't had an issue with it so I can only assume it is a python 2 issue. As we are heading into Kodi 21 in the next couple months it's going to be difficult finding any devs that support Kodi 18.9. I can ask the dev to see if there is any kind of fix that could be done, but this add-on could be end-of-life for anything less than LibreELEC 10.0.

**Comment by mrgsi** [comment]
The Developer is looking into the issue and will try to fix it. In the mean time you can fall back to version 11.6 as I placed it in an archive folder here under the WNT2 repository. Just disable the auto update, download the zip file onto your device, then install from zip. That version does work, it just blocks you from watching some movies. All TV shows should still work.

**Comment by mrgsi** [comment]
New version 12.0 (no longer beta) is now available and should fix the python 2 issue.

**Comment by pitsi** [comment]
Fixed with 0.12 final. Thank you so much!
Please tell me where is the option to clean its cache, if you know where it is.

And an apology about why I am still on kodi 18. 
My "htpc" is an old laptop with a dual core cpu and an intel vga which can not do hardware acceleration. Despite that, it can play low bitrate 720p files just fine, on le 9.2.x and kodi 18.

On le 10 however, kodi 19 has framedrops even on 480p. Framedrops equivalent to playing a 1080p file on that same hw, that get annoying after a while and I still haven't found out why. I first tried it when it was at its first beta in early 2021, did all the major updates until 10.0.4 last January, and then the 2-3 nightlies it got until the end of the same month, which were the very last ones for it. The issue remained, so I decided to stay on le 9 and wait to see how it will behave 11.

When the first beta of le 11 was relesed, I tried it right away. Le 11 still has the same framedrops and it brought one more serious (imho) issue: wifi issues due to some... incompatibility between iwd and connman. Iwd replaced wpa_supplicant on le 11.
I tried ALL its nightlies from late May to early September of 2022 on my rpi3b+ and I am really sad to see the issue reaching stable and the devs saying "connect via ethernet". Thus, no le 11 for me too, with one more reason to avoid it.
The problem is that unlike the rpi which has a couple of kodi-specific distros one can install, e.g. osmc, there is only le for x64, so I have no other thing to try there.

Last but not least, I have an amlogic s805 (amlmx) for my nieces. It has a le 9.2.8.x build by dtech (sole maintainer for it) and it works just fine with it. The only newer le version for it was some pre-pre-alpha of le 10, so let's say that le's development has stopped to le 9.2.x and kodi 18 for it. It is exactly the same situation as coreelec legacy in fact.
Tbh, taying on that version of le on my laptop too helps troubleshoot stuff that are related to that old version of kodi and python2 on their end. It saves me travelling ~20km everytime they spot an issue on it, e.g. youtube's throttling issue which was patched 1+ year ago.

Thank you for taking some time to read all of it :)
